python/lsst/cell_coadds/_grid_container.py

```diff
@@ -0,0 +1,201 @@
+                f"{self.offset.x + self.shape.x -1}."
+            )
+        if index.y < self.offset.y or index.y >= self.offset.y + self.shape.y:
+            raise ValueError(
+                f"y index {index.y} out of range; expected a value between {self.offset.y} and "
+                f"{self.offset.y + self.shape.y -1}."
+            )
+
+    def __repr__(self) -> str:
+        return f"{type(self).__name__}(shape={self.shape}, offset={self.offset})"
+
+    # Implement the necessary methods for MutableMapping.
+    def __len__(self) -> int:
+        return len(self._cells)
+
+    def __getitem__(self, index: Index2D) -> T:
+        return self._cells[index]
+
+    # def __contains__(self, index: object) -> bool:
+    #     return index in self._cells
+
+    def __setitem__(self, index: Index2D, value: T) -> None:
+        self._check(index)
+        self._cells[index] = value
+
+    def __delitem__(self, index: Index2D) -> None:
+        self._check(index)
+        del self._cells[index]
+
+    def __iter__(self) -> Iterator[T]:
+        yield from self._cells.values()
+
+    def keys(self) -> Iterable[Index2D]:  # populated_indices
+        """Return an iterator over the indices in the container with values.
+
+        See also
+        --------
+        indices
+        """
+        yield from self._cells.keys()
+
+    def indices(self) -> Iterable[Index2D]:
+        """Return an iterator over all possible indices for the container.
+
+        Unlike `keys`, this method returns an iterator over all valid indices,
+        whether the corresponding value is set or not.
+
+        See also
+        --------
+        keys
+        """
+        iterator = product(
+            range(self.offset.y, self.offset.y + self.shape.y),
+            range(self.offset.x, self.offset.x + self.shape.x),
+        )
+        for y, x in iterator:
+            yield Index2D(x, y)
+
+    def get(self, index: Index2D, default: T | None = None) -> T | None:
+        """Return the value at the given index, or a default value if not set.
+
+        Parameters
+        ----------
+        index : `~lsst.skymap.Index2D`
+            Index of the cell to return.
+        default : T
+            Value to return if the cell is not set.
+
+        Returns
+        -------
+        value : T
+            The value at the given index, or the default value if not set.
+        """
+        return self._cells.get(index, default)
+
+    @property
+    def shape(self) -> Index2D:
+        """Number of cells in the container in each dimension."""
+        return self._shape
+
+    @property
+    def offset(self) -> Index2D:
+        """Index of the first cell in the container."""
+        return self._offset
+
+    @property
+    def size(self) -> int:
+        """Return the number of cells expected in the container.
+        This does not indicated the number of cells that have been filled.
```

Comment:
Add newline after first sentence, and just say, "The number" instead of "Return the number" since this is attribute-like instead of method-like.